lebedigital/calibration/utils.py

```diff
@@ -0,0 +1,238 @@
+    Returns:
+
+    """
+    """
+    Queries KG, the updated one done by Illias (I think from KG team). Lot of funky words as I dont know what they are.
+    Returns:
+
+    """
+    if KG:
+
+        def query_objects(queries, graph):
+            # function to get objects from specific subject, predicate pairs, knowing there is only one result
+            results = {}
+            for query in queries:
+                q = f"""
+                                SELECT ?object
+                                WHERE {{
+                                        {queries[query]['subject']}  {queries[query]['predicate']} ?object
+                                }}
+                        """
+
+                result = graph.query(q)
+                for r in result:
+                    results[query] = r["object"]
+
+            return results
+
+        # define queries
+        queries = {
+            "diameter": {
+                "subject": "ns1:informationbearingentity1",
+                "predicate": "ns1:has_decimal_value",
+            },
+            "length": {
+                "subject": "ns1:informationbearingentity2",
+                "predicate": "ns1:has_decimal_value",
+            },
+            "path": {
+                "subject": "ns1:informationbearingentity9",
+                "predicate": "ns9:has_text_value",
+            },
+            "file_name": {
+                "subject": "ns1:informationbearingentity8",
+                "predicate": "ns9:has_text_value",
+            },
+        }
+
+        # Path to KG, for testing
+        # path_to_KG = '../../usecases/MinimumWorkingExample/emodul/knowledge_graphs/BA-Losert MI E-Modul 28d v. 04.08.14 Probe 4.ttl'
+
+        path_to_KG = os.path.join(path, exp_name)
+        # initialize the graph
+        knowledge_graph = rdflib.Graph()
+        knowledge_graph.parse(path_to_KG, format="ttl")
+
+        # queries
+        results = query_objects(queries, knowledge_graph)
+
+        # read csv into dataframe
+        file_path = results["path"] + "/" + results["file_name"]
+    # skip_last = 145 # hard coded to get the third loading cycle. need somthing better
+    # skip_init = 330
+    if KG == False:  # this is just for testing purposes
+        file_path = os.path.join(path, exp_name)
+        results = {}
+        results["length"] = 300.2
+        results["diameter"] = 98.6
+    df = extract_third_load_cycle(data_path=file_path)
+    df["displacement"] = (
+        df["Transducer 1[mm]"] + df["Transducer 2[mm]"] + df["Transducer 3[mm]"]
+    ) / 3
+    # df['stress'] = df['Force [kN]'] / (np.pi * (float(results['diameter']) / 2) ** 2)
+
+    output = {
+        "exp_name": exp_name,
+        "height": float(results["length"]),
+        "diameter": float(results["diameter"]),
+        "displacement": np.array(df["displacement"]),
+        "force": np.array(df["Force [kN]"]),
+    }
+    return output
+
+
+class PosteriorPredictive:
```

Comment:
I would not expect this class definition to be in "utils.py", I would suggest to have clear/more thought through structure for the script that will allow for extension to the demonstrator

Reply:
Yes good point. In demonstrator this can be/will be used as follows: After we get x^* (the optimized design variable), we would get the prediction of the KPIs +- sd. Essentially running your snakemake several times, for different sample of (hydration parameters and paste properties E and strength).